File: modules/raft/src/main/java/org/apache/ignite/raft/jraft/storage/snapshot/remote/Session.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.ignite.raft.jraft.storage.snapshot.remote;

import java.io.Closeable;
import org.apache.ignite.raft.jraft.Status;

/**
 * A copy session.
 */
public interface Session extends Closeable {

    /**
     * Cancel the copy job.
     */
    void cancel();

    /**
     * Block the thread to wait the copy job finishes or canceled.
     *
     * @throws InterruptedException if the current thread is interrupted while waiting
     */
    void join() throws InterruptedException;

    /**
     * Returns the copy job status.
     */
    Status status();
}

File: ignite-extensions/modules/ignite-graph/works-ignite-streamer/src/main/scala/de/kp/works/ignite/core/Session.scala
package de.kp.works.ignite.core

/**
 * Copyright (c) 2021 - 2022 Dr. Krusche & Partner PartG. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 *
 * @author Stefan Krusche, Dr. Krusche & Partner PartG
 *
 */

import de.kp.works.ignite.conf.WorksConf
import org.apache.spark.sql.SparkSession

object Session {

  private val session: SparkSession = initialize

  def initialize: SparkSession = {
    /*
     * Initialize the overall configuration
     * with the internal configuration file
     */
    WorksConf.init()
    /*
     * Extract internal Spark configuration
     */
    val sparkCfg = WorksConf.getSparkCfg
    val driverCfg = sparkCfg.getConfig("driver")

    val driver_maxResultSize = driverCfg.getString("maxResultSize")
    val driver_memory = driverCfg.getString("memory")

    val executorCfg = sparkCfg.getConfig("executor")

    val executor_instances = executorCfg.getString("instances")
    val executor_memory = executorCfg.getString("memory")

    val spark: SparkSession = SparkSession.builder()
      .appName(sparkCfg.getString("appName"))
      .master(sparkCfg.getString("master"))
      /*
       * Driver & executor configuration
       */
      .config("spark.driver.maxResultSize", driver_maxResultSize)
      .config("spark.driver.memory", driver_memory)
      .config("spark.executor.memory", executor_memory)
      .config("spark.executor.instances", executor_instances)
      .getOrCreate()

    spark.sparkContext.setLogLevel("ERROR")
    spark

  }

  def getSession: SparkSession = session

}
